students/N0vA/lesson_05/mailroom3.py

```diff
@@ -0,0 +1,133 @@
+                    'We appreciate your support not only for us but for our cause.\n\n'
+                    'We wish you all the best,\n\n'
+                    'Local Charity Persident\n')
+
+
+# Define sort key
+def sort_key(donor):
+    return int(sum(donor[1]))
+
+# Create a report of donors
+def report():
+
+    # Sort Data
+    sorted_data = sorted(database.items(), key=sort_key, reverse=True)
+
+    # Format table
+    member_row = '{:<24}{:^5} ${:>14,}{:^5} {:^5}{:^5} ${:>14,.2f}'
+    print('Generating report of donors....')
+    # Header
+    print(""+"-" * 80 + "\n Donor Name"+" " * 19 + "| Total Donated | Num Donations | Average Donation\n"+"-" * 80)
+
+    # Print 
+    for per in sorted_data:
+        print(member_row.format(per[0], ' ', int(sum(per[1])), ' ', round(len(per[1]),2), ' ', round(sum(per[1])/len(per[1]),2)))
+    
+    # Exit to main menu
+    exit = 'none'
+    while exit != 'quit':
+        exit = input('Type quit to return to the menu... ')
+
+def quick_letter():
+    # Write a letter to all donors
+
+    print('Ok.  Sending letters to all donors now...')
+
+    pth = pathlib.Path('./')
+    folder = pth.absolute()
+
+    for donor in database.keys():
+        first_name = donor.split()[0]
+        last_name = donor.split()[1]
+        file_format = "{0}_{1}.txt".format(first_name, last_name)
+        file_path = folder / file_format
+
+        # Donation amount
+        charity_amount = sum(database[donor])
+
+        # Write new files
+        with open(file_path, 'w') as new_file:
+            letter = ('Dear {0}\n\n'
+                'Thank you for your donations totaling' 
+                '$ {1:,}.  We appreciate your contributions'
+                'for the year.\n\nHappy holidays,\n\n' 
+                'Your loal charity President').format(first_name, charity_amount)
+
+            # Writing files
+            new_file.write(letter)
+
+       
+# Execute file when running
+if __name__ == '__main__':
+    
+    arg_dict = {1: thank_you,
+                2: report,
+                3: quick_letter,
+                4: exit}
+
+    while True:
+    # Opens up the mailroom
+        task = 0
+        task = input("\n".join(("What do you need to do?",
+              "Please choose from the options below:",
+              "1 - Send Thank You Card",
+              "2 - Print A Report",
+              "3 - Send Thank You to all donors",
+              "4 - Exit",
+              ">>> ")))
+
+        # Run functions for tasks based on user's response
+        try:
+            task = int(task)
+        except ValueError:
+            task = int(input('Please enter a valid option from 1-4: '))
+        else:
+            # I tried except KeyError here and that didn't work
+            if task not in arg_dict.keys():
+                task = int(input('Please enter a valid option from 1-4: '))
+        arg_dict.get(task)()
```

Comment:
if you used `arg_dict[task]` then you could catch KeyError

using `arg_dict.get()` will never throw a KeyError, even if key does not exist. It can throw a TypeError though, if key does not exist then `get()` will return `None` and executing `None()` will return `TypeError`